onnxruntime/core/providers/dml/DmlExecutionProvider/src/GraphDescBuilder.cpp

```diff
@@ -31,6 +32,121 @@ namespace Dml::GraphDescBuilder
+            NodeData* node = &nodesData[nodeIndex];
+
+            if (node->state == NodeState::Visited)
+            {
+                nodeIndicesToVisit.pop();
+                continue;
+            }
+
+            if (node->state == NodeState::Visiting)
+            {
+                // The stack has been popped all the way back to this node, which means all its predecessors have been
+                // visited. That means we're done visiting this node too.
+                node->state = NodeState::Visited;
+                nodeIndicesToVisit.pop();
+                continue;
+            }
+
+            node->state = NodeState::Visiting;
+
+            for (uint32_t predecessorNodeIndex : node->predecessorIndices)
+            {
+                // If we're already visiting that node, we are in a cycle and we should fail early
+                ORT_THROW_HR_IF(E_INVALIDARG, nodesData[predecessorNodeIndex].state == NodeState::Visiting);
+                nodeIndicesToVisit.push(predecessorNodeIndex);
+            }
+        }
+
+        // Collect all the unvisited nodes
+        std::unordered_set<uint32_t> orphanNodes;
+        for (uint32_t nodeIndex = 0; nodeIndex < nodesData.size(); ++nodeIndex)
+        {
+            if (nodesData[nodeIndex].state == NodeState::NotVisited)
+            {
+                orphanNodes.insert(nodeIndex);
+            }
+        }
+
+        // Mapping from the old indices to the new indices that have been shifted after removing earlier nodes
+        std::vector<uint32_t> shiftedIndicesMapping(graphNodes.size());
+
+        uint32_t shift = 0;
+        for (uint32_t nodeIndex = 0; nodeIndex < graphNodes.size(); ++nodeIndex)
+        {
+            if (nodesData[nodeIndex].state == NodeState::NotVisited)
+            {
+                // The node is not connected, so we simply increase the shift value (the node will be overwritten by the following nodes)
+                ++shift;
+            }
+            else
+            {
+                // The node is connected, so we keep it and adjust its mapping
+                shiftedIndicesMapping[nodeIndex] = nodeIndex;
+                graphNodes[nodeIndex - shift] = std::move(graphNodes[nodeIndex]);
+                shiftedIndicesMapping[nodeIndex] = nodeIndex - shift;
+            }
+        }
+
+        graphNodes.resize(graphNodes.size() - shift);
+
+        // Adjust the node indices in the input edges
+        for (auto& inputEdge : graphInputEdges)
+        {
+            inputEdge.ToNodeIndex = shiftedIndicesMapping[inputEdge.ToNodeIndex];
+        }
+
+        // Adjust the node indices in the output edges
+        for (auto& outputEdge : graphOutputEdges)
+        {
+            outputEdge.FromNodeIndex = shiftedIndicesMapping[outputEdge.FromNodeIndex];
+        }
+
+        // Adjust the node indices in the intermediate edges
+        for (auto& intermediateEdge : graphIntermediateEdges)
```

Comment:
We should also remove _orphan_ (edge which feeds to the orphan/removed node) intermediate edges from the `graphIntermediateEdges`. Also, for orphan nodes, `shiftedIndicesMapping` will have default value (may be 0). So _orphan_ intermediate edges will point to nodeIndex 0.